### DISCUSSION TITLE
Azure Dev Ops - Error: Unknown command 'Publish'

### DISCUSSION THREAD

**Discussion by cstanleyhns** [cake-build/cake#3174] [Q&A] [answered]
### What You Are Seeing?
We have a number of .Net based projects that use a Cake build in AZure Dev Ops pipelines. The process is failing during the project build phase with the following error:

```
2021-02-17T15:12:29.8490934Z Executing build script...
2021-02-17T15:12:29.9209087Z ##[command]"D:\a\1\s\tools\Cake\Cake.exe" "D:\a\1\s\build.cake" -target="Publish" -verbosity="Verbose" --paths_tools="D:\a\1\s\tools" -outputDirectory=D:\a\1\a
2021-02-17T15:12:30.3673785Z 
2021-02-17T15:12:30.3674635Z Error: Unknown command 'Publish'.
2021-02-17T15:12:30.3674922Z 
2021-02-17T15:12:30.3675385Z        D:\a\1\s\build.cake -target=Publish -verbosity=Verbose 
2021-02-17T15:12:30.3680682Z --paths_tools=D:\a\1\s\tools -outputDirectory=D:\a\1\a
2021-02-17T15:12:30.3681421Z                                    ^^^^^^^ No such command
2021-02-17T15:12:30.3681705Z 
2021-02-17T15:12:30.6211384Z ##[error]Process 'Cake.exe' exited with code '-1'.
2021-02-17T15:12:30.6965233Z ##[section]Finishing: Build

```
**Azure Pipelines YML:**
```
trigger:
- master
resources:
- repo: self
queue:
  name: Hosted VS2017
steps:
- task: UseDotNet@2
  inputs:
    packageType: 'sdk'
    version: '2.2.x'
- task: UseDotNet@2
  inputs:
    packageType: 'sdk'
    version: '3.1.x'
- task: CacheBeta@0
  displayName: 'Nuget cache'
  inputs:
    key: 'crm-api-nuget-package-cache'     
    path: '$(UserProfile)\.nuget\packages'
- task: cake-build.cake.cake-build-task.Cake@0
  displayName: Build
  inputs:
    target: Publish
    arguments: '-outputDirectory=$(Build.ArtifactStagingDirectory)'
- task: PublishBuildArtifacts@1
  displayName: 'Publish Artifact: drop'
```

Full logs here: 
Log: [cake log.zip](https://github.com/cake-build/cake/files/6007229/cake.log.zip)
Copy of YML file: [azure-pipelines.zip](https://github.com/cake-build/cake/files/6007202/azure-pipelines.zip)
Cake File: [build-cake.zip](https://github.com/cake-build/cake/files/6007217/build-cake.zip)

This has just started happening with no codebase changes to the projects. Build.cake files have not changed and azure-pipelines.yml files have not changed.

I raised a ticket with Microsoft but as Cake is a thrd party, they directed me here.

I have gone through the configuration files and cannot see any changes that have been made to cause this so would really appreciate some help and insight in have to uncover what is wrong.

### What is Expected?
Normal operation is the cake build works fine and that the resulting artifacts are published which then triggers the Relase pipeline to run.

### What version of Cake are you using?
From the logs, it looks like Cake version is 0.4.4 although there is a installation form nuget of cake1.0.0

### Are you running on a CI Server?  If so, which one?

The cake process is running within Azure Dev Ops

### How Did You Get This To Happen? (Steps to Reproduce)

Any project project that is using cake is expereincing this issue.

**Comment by augustoproiete** [comment, marked as answer]
@cstanleyhns It looks like you never fixed/pinned your Cake version and have been living on the edge for a very long time (~2015 if version 0.4.x is correct :sweat_smile:), and because of that your build processes have always been using the latest version of Cake available at the time your builds run - and you've been lucky enough that it didn't break sooner! :four_leaf_clover:

[We released Cake v1.0 a couple of weeks ago](https://cakebuild.net/blog/2021/02/cake-v1.0.0-released), which includes a number of new features and also breaking changes, which seems to be what's affecting you.

Your options at this point would be:

**(1)** [Follow the instructions on the Reproducible Builds](https://cakebuild.net/docs/writing-builds/reproducible-builds/) section and [pin the Cake version](https://cakebuild.net/docs/writing-builds/reproducible-builds/pinning-cake-version) to version `0.38.5`, the release prior to `1.0.0` which you would have been using without issues for the past 5 months. That's probably the easiest solution and should get you back to where you were two weeks ago.

or

**(2)** [Migrate your builds to Cake v1.0](https://cakebuild.net/docs/getting-started/upgrade#cake-cli-updates) making the necessary changes to your bootstrappers and possibly updating addins/modules that you use to newer versions compatible with Cake v1.0.

---

Regardless of which way you go, make sure you [pin the Cake version](https://cakebuild.net/docs/writing-builds/reproducible-builds/pinning-cake-version) version of Cake moving forward so that when we release Cake `2.0.0` this doesn't happen again - and you can move to `2.0.0` at your own pace.

Other posts that might be useful (similar to yours):

- [How to migrate to V1.0.0 - CLI not working?](https://github.com/cake-build/cake/discussions/3100)
- [Cake build is not working #3110](https://github.com/cake-build/cake/discussions/3110)

**Reply by cstanleyhns** [reply]
Hi, really appreciate your quick response and guidence. I will try and progress this today and let you know if i have further issues.

Chris

**Reply by cstanleyhns** [reply]
Hi,

The packages.config file was pinned to v0.3.8.1, changed to v0.3.8.5 but then noticed that the Tools folder that contains this file was not part of source control and therefore not in the repos. Added it and then the builds work fine.

Thanks again for your guidance.

Chris
